brushfire-core/src/main/scala/com/stripe/brushfire/TreeTraversalStrategy.scala

```diff
@@ -0,0 +1,149 @@
+   * @return the leaf node that best matches row
+   */
+  def find(init: Node[K, V, T, A], row: Map[K, V]): Option[LeafNode[K, V, T, A]]
+
+  /**
+   * Walks down the tree, following the true predicates to find the leaf that
+   * contains `row`. At each `SplitNode`, this will find all the `true`
+   * predicates and call `choose` to resolve these down to at-most 1 node to
+   * follow.
+   *
+   * @param init   the initial node to start the walk at
+   * @param row    the "row"/instance we're trying to classify
+   * @param choose a function that chooses at-most 1 node from a list of candidates
+   * @return the leaf node that best matches row
+   */
+  protected def findMatching(init: Node[K, V, T, A], row: Map[K, V])(choose: ((Node[K, V, T, A], List[Node[K, V, T, A]])) => Option[Node[K, V, T, A]]): Option[LeafNode[K, V, T, A]] = {
+    def recur(node: Node[K, V, T, A]): Option[LeafNode[K, V, T, A]] = {
+      node match {
+        case leaf @ LeafNode(_, _, _) =>
+          Some(leaf)
+
+        case SplitNode(children) =>
+          val candidates = children.collect {
+            case (key, pred, child) if pred(row.get(key)) => child
+          }.toList
+          choose(node -> candidates).flatMap(recur)
+      }
+    }
+
+    recur(init)
+  }
+}
+
+object TraversalStrategy {
+
+  /** Returns an instance of [[FirstMatch]]. */
+  def firstMatch[K, V, T, A]: FirstMatch[K, V, T, A] =
+    new FirstMatch[K, V, T, A]
+
+  /** Returns an instance of [[ProbabilisticWeightedMatch]] whose seed is `None`. */
+  def probabilisticWeightedMatch[K, V, T, A](implicit toDouble: A => Double): ProbabilisticWeightedMatch[K, V, T, A] =
+    new ProbabilisticWeightedMatch[K, V, T, A](None)
+
+  /** Returns an instance of [[MaxWeightedMatch]]. */
+  def maxWeightedMatch[K, V, T, A: Ordering]: MaxWeightedMatch[K, V, T, A] =
+    new MaxWeightedMatch[K, V, T, A]
+}
+
+/**
+ * A [[TraversalStrategy]] that will always choose the first matching candidate
+ * in a split.
+ */
+final class FirstMatch[K, V, T, A] extends TraversalStrategy[K, V, T, A] {
+  def find(init: Node[K, V, T, A], row: Map[K, V]): Option[LeafNode[K, V, T, A]] =
+    findMatching(init, row)(_._2.headOption)
+}
+
+/**
+ * A [[TraversalStrategy]] that will choose a path probabilistically, weighting
+ * the choices by using the annotations as a weight. So, if there are 2
+ * candidate paths, `a` and `b`, and `a.annotation` is twice the weight of
+ * `b.annotation`, then `a` is twice as likely to be chosen (though `b` may
+ * still be chosen!).
+ *
+ * The actual path chosen depends either on `seed`, if one is provided, or the
+ * `row` itself if no `seed` is provided. So, given the same `row`, this should
+ * always result in the same leaf being found.
+ *
+ * @param seed an optional seed for the RNG used in a single traversal
+ */
+final class ProbabilisticWeightedMatch[K, V, T, A](seed: Option[Long] = None)(implicit toDouble: A => Double) extends TraversalStrategy[K, V, T, A] {
```

Comment:
If only we had some nice, sane `Hash` type class ;) Agreed though - I'll make the seed required. I'm a bit worried about how far to leak this in our use case, but now that I think of it, it will probably pay off in the end to expose it pretty far down.